modules/041-linstor/hooks/metrics.go

```diff
@@ -0,0 +1,303 @@
+
+	return linstorResourceSnapshot{
+		IsSnapshot: snapshotName != "",
+	}, nil
+}
+
+// collectLinstorMetrics
+//
+// synopsis:
+//   Waits for internal LINSTOR resources be installed then
+//   collects general statistics from them
+//
+func collectLinstorMetrics(input *go_hook.HookInput) error {
+	input.MetricsCollector.Set(telemetry.WrapName("linstor_enabled"), 1, map[string]string{})
+
+	// LINSTOR manages it's own internal CRDs, so we need to wait for them before starting the watch
+	if linstorMetricsHookConfig.Kubernetes[0].Kind == "" {
+
+		var version string
+		for _, sRaw := range input.Snapshots[linstorCRDsSnapshot] {
+			discoveredVersion := sRaw.(string)
+			if version == "" {
+				version = discoveredVersion
+			} else if version != discoveredVersion {
+				return fmt.Errorf("LINSTOR internal CRDs have different storage versions")
+			}
+		}
+
+		if len(input.Snapshots[linstorCRDsSnapshot]) >= 4 {
+			// LINSTOR installed
+			input.LogEntry.Infof("LINSTOR internal CRDs installed, update kind for binding linstor resources to collect metrics")
+			apiVersion := "internal.linstor.linbit.com/" + version
+			*input.BindingActions = append(*input.BindingActions, go_hook.BindingAction{
+				Name:       linstorNodesSnapshot,
+				Action:     "UpdateKind",
+				ApiVersion: apiVersion,
+				Kind:       "Nodes",
+			})
+			*input.BindingActions = append(*input.BindingActions, go_hook.BindingAction{
+				Name:       linstorSPsSnapshot,
+				Action:     "UpdateKind",
+				ApiVersion: apiVersion,
+				Kind:       "NodeStorPool",
+			})
+			*input.BindingActions = append(*input.BindingActions, go_hook.BindingAction{
+				Name:       linstorRDsSnapshot,
+				Action:     "UpdateKind",
+				ApiVersion: apiVersion,
+				Kind:       "ResourceDefinitions",
+			})
+			*input.BindingActions = append(*input.BindingActions, go_hook.BindingAction{
+				Name:       linstorResourcesSnapshot,
+				Action:     "UpdateKind",
+				ApiVersion: apiVersion,
+				Kind:       "Resources",
+			})
+			// Save new kind as current kind.
+			linstorMetricsHookConfig.Kubernetes[0].Kind = "Nodes"
+			linstorMetricsHookConfig.Kubernetes[0].ApiVersion = apiVersion
+			linstorMetricsHookConfig.Kubernetes[1].Kind = "NodeStorPool"
+			linstorMetricsHookConfig.Kubernetes[1].ApiVersion = apiVersion
+			linstorMetricsHookConfig.Kubernetes[2].Kind = "ResourceDefinitions"
+			linstorMetricsHookConfig.Kubernetes[2].ApiVersion = apiVersion
+			linstorMetricsHookConfig.Kubernetes[3].Kind = "Resources"
+			linstorMetricsHookConfig.Kubernetes[3].ApiVersion = apiVersion
+			// Binding changed, hook will be restarted with new objects in snapshot.
+			return nil
+		}
+		// LINSTOR is not yet installed, do nothing
+		return nil
+	}
+
+	// Start main hook logic
+
+	// Collect general amount of lisntor nodes
+	var linstorSatellitesCount float64
+	for _, sRaw := range input.Snapshots[linstorNodesSnapshot] {
+		s := sRaw.(linstorNodeSnapshot)
+
+		// NodeType:
+		// 1 - controller
```

Comment:
Let add constant with types instead comment in the code

Reply:
Good idea, done